[mlir] Extract RHS rows once when lowering vector.contract to dot

abulavin opened the pull request:
The `vector.contract` op on two matrices A and B will be lowered to individual dot products of each row and column of A and B respectively. The existing lowering will extract each column of B for each row of A, which leads to multiple values in the IR representing the same columns of B.

This PR makes changes to the `ContractOpToDotLowering` to make sure that the columns of B are only ever extracted once, so then the SSA values representing the extracted columns are then re-used in the IR for later dot products.

I have updated the existing vector-contract-to-dot-transforms test.

abulavin committed:
Extract RHS rows once when lowering vector.contract to dot

banach-space commented on mlir/lib/Dialect/Vector/Transforms/LowerVectorContract.cpp:
```diff
@@ -758,12 +758,20 @@ FailureOr<Value> ContractionOpToDotLowering::matchAndRewriteMaskableOp(
-                    ? rhs
-                    : rewriter.create<vector::ExtractOp>(op.getLoc(), rhs, c);
+      if (r == 0) {
+        // We only need to extract the rows of the RHS once
+        // and then re-use them later.
```
This comment refers to rows while below there's `extractedCols`

banach-space commented on mlir/test/Dialect/Vector/vector-contract-to-dot-transforms.mlir:
```diff
@@ -169,13 +169,11 @@ func.func @extract_contract3(%arg0: vector<3xf32>,
 // CHECK:    %[[T23:.*]] = vector.extract %[[A]][1] : vector<2xf32> from vector<2x2xf32>
-// CHECK:    %[[T24:.*]] = vector.extract %[[Bt]][0] : vector<2xf32> from vector<2x2xf32>
-// CHECK:    %[[T32:.*]] = arith.mulf %[[T23]], %[[T24]] : vector<2xf32>
+// CHECK:    %[[T32:.*]] = arith.mulf %[[T23]], %[[T2]] : vector<2xf32>
```
With the new [Testing Guide for MLIR ](https://github.com/llvm/mlir-www/pull/216) in mind (soon to be merged), could you replace `T32` and `T41` with some meaningful names? Right now it's impossible to tell that these are columns of `B`.

Bonus points for updating all CHECK lines.

abulavin replied:
The guide is helpful, thank you.

banach-space reviewed (commented):
Thanks, mostly just nits.

abulavin committed:
fixup: Refactor variable names in vector contract to dot transform test

abulavin committed:
fixup: refer to columns instead of rows in reuse comment

banach-space commented on mlir/lib/Dialect/Vector/Transforms/LowerVectorContract.cpp:
```diff
@@ -758,12 +758,20 @@ FailureOr<Value> ContractionOpToDotLowering::matchAndRewriteMaskableOp(
+                      : rewriter.create<vector::ExtractOp>(op.getLoc(), rhs, c);
+        extractedCols.push_back(b);
+      }
+      Value b = extractedCols[c];
```
[nit] Avoid re-using/re-defining variable names. Btw, you could use `lhs` and `rhs` for `a` and `b`, respectively.

banach-space commented on mlir/lib/Dialect/Vector/Transforms/LowerVectorContract.cpp:
```diff
@@ -758,12 +758,20 @@ FailureOr<Value> ContractionOpToDotLowering::matchAndRewriteMaskableOp(
-                    ? rhs
-                    : rewriter.create<vector::ExtractOp>(op.getLoc(), rhs, c);
+      if (r == 0) {
+        // We only need to extract the columns of the RHS once
+        // and then re-use them later.
```
[nit] This implies that RHS is somehow special, but that's not really what you mean, right? Instead, you are making sure that for _both_ LHS and RHS we extract the rows/cols only once. Could you rephrase?

banach-space reviewed (approved):
I love the updates in the test, thank you!

LGTM % nits

abulavin committed:
fixup: Rephrase comment and rename column/row variables

abulavin committed:
fixup: remove trailing whitespace

abulavin commented:
@banach-space Hi again, I believe I've addressed all your comments, so if there's nothing more then could this please be merged for me? I can see that the `Check code formatting` check failed after my last commit to remove some trailing whitespace that got caught by this check. The most recent check failure actually has nothing to do with code formatting and looks like an infra issue. Can we just ignore this or is there something I can do?

banach-space commented:
> @banach-space Hi again, I believe I've addressed all your comments, so if there's nothing more then could this please be merged for me? I can see that the `Check code formatting` check failed after my last commit to remove some trailing whitespace that got caught by this check. The most recent check failure actually has nothing to do with code formatting and looks like an infra issue. Can we just ignore this or is there something I can do?

Thanks for the ping! Sorry, I didn't realise those failures were bogus. Yes, I think that it is safe to land this.
